File: solutions/python/simple-cipher/1/simple_cipher.py
# Tribute to my Shree DR.MDD

import itertools as it
import random as rnd
import string as st

alpha_base = ord("a")


def mutate(x: str, y: str, inverse: bool) -> str:
    p = ord(x) - alpha_base
    q = ord(y) - alpha_base
    if inverse:
        q = -q
    z = (p + q) % 26
    return chr(z + alpha_base)


class Cipher:
    def __init__(self, secret: str | None = None):
        if secret is None:
            secret = "".join([rnd.choice(st.ascii_lowercase) for _ in range(100)])
        self.key = secret

    def encode(self, data: str, inverse: bool = False) -> str:
        return "".join([
            mutate(a, b, inverse)
            for a, b in zip(data, it.cycle(self.key))
        ])

    def decode(self, encrypted: str) -> str:
        return self.encode(encrypted, inverse=True)

File: solutions/elixir/simple-cipher/1/lib/simple_cipher.ex
# For Shree DR.MDD

defmodule SimpleCipher do
  def encode(plain_chr, key_chr) when is_integer(plain_chr) do
    shifted = plain_chr - ?a + key_chr
    if shifted <= ?z, do: shifted, else: ?a + (shifted - ?z - 1)
  end

  def decode(cipher_chr, key_chr) when is_integer(cipher_chr) do
    shifted = cipher_chr - (key_chr - ?a)
    if shifted >= ?a, do: shifted, else: ?z - (?a - shifted) + 1
  end

  def cipher(text, key, fun) do
    {text_list, key_list} = {String.to_charlist(text), String.to_charlist(key)}

    Stream.zip(text_list, Stream.cycle(key_list))
    |> Enum.map(fn {t_chr, k_chr} -> fun.(t_chr, k_chr) end)
    |> List.to_string()
  end

  def encode(plaintext, key), do: cipher(plaintext, key, &encode/2)

  def decode(ciphertext, key), do: cipher(ciphertext, key, &decode/2)

  def generate_key(len) do
    for _ <- 1..len, into: "", do: <<Enum.random(?a..?z)>>
  end
end
